### PR TITLE
feat(Core/ItemHandler): Optional item recovery

### PR DESCRIPTION
**Pull request by IntelligentQuantum**
It is for the [AzerothJS](https://github.com/AzerothJS/AzerothAPI)

### CODE REVIEW

**Comment by FrancescoBorzi** [comment]
hey @IntelligentQuantum thanks for the PR

> It is for the AzerothJS

can you elaborate more? what features does this PR implements and in which way are they required to be in the core?

**Comment by Helias** [comment]
Features of this PR:
- it allows to have a log about deleted items
- It allows to recover items via SQL

This PR adds 3 configs:
- CONFIG_ITEMDELETE_METHOD
- CONFIG_ITEMELETE_QUALITY
- CONFIG_ITEMELETE_ITEM_LEVEL

These configs allow to choose the method, quality and item_level of the items that could be logged in the table `recovery_item`.

Do we need this?
Yes, because this is an extra-feature of AC. 
Besides, I don't think that this bother the blizzlike game, indeed because it's a config, so you can enable/disable.
Moreover, we can implement third-party APIs, modules or tools that work with AC.

Here a video made by @IntelligentQuantum that shows how this feature works (using the AzerothJS API).

https://mega.nz/#!pyYxlYyJ!vy6-5kGcBiTmE6xOOpbBWPuSAZvhhi2jGiFo4FXXHfM

**Comment by FrancescoBorzi** [comment]
thanks @Helias for the description and @IntelligentQuantum let us know when the PR is ready to be reviewed

**Comment by IntelligentQuantum** [comment]
Hi @FrancescoBorzi it's ready

**Comment by Ali-Khazaee** [comment]
As i used this feature before, There is an issue when you sell items to vendors there will be no log

**Comment by MarkMBarnett** [comment]
Tested successfully

File: src/server/worldserver/worldserver.conf.dist [hunk 1]
```diff
@@ -2957,6 +2958,43 @@ CharDelete.KeepDays = 30
...
+#        Default:     0 - (Completely remove item from the database)
+#                     1 - (Save Item to database)
+
+ItemDelete.Method = 0
+
+#
+#    ItemDelete.Vendor
+#        Description:
+#        Default:     0
+#
```
Comment (FrancescoBorzi) [inline src/server/worldserver/worldserver.conf.dist]: missing description

File: src/server/worldserver/worldserver.conf.dist [hunk 2]
```diff
@@ -2957,6 +2958,43 @@ CharDelete.KeepDays = 30
...
+
+ItemDelete.Vendor = 0
+
+#
+#    ItemDelete.Quality
+#        Description:
+#        Default:     3
```
Comment (FrancescoBorzi) [inline src/server/worldserver/worldserver.conf.dist]: missing description

File: src/server/worldserver/worldserver.conf.dist [hunk 3]
```diff
@@ -2957,6 +2958,43 @@ CharDelete.KeepDays = 30
...
+#
+
+ItemDelete.Quality = 3
+
+#
+#    ItemDelete.ItemLevel
+#        Description:
+#        Default:     80
+#
```
Comment (FrancescoBorzi) [inline src/server/worldserver/worldserver.conf.dist]: missing description

File: src/server/worldserver/worldserver.conf.dist [hunk 4]
```diff
@@ -2970,23 +2970,35 @@ ItemDelete.Method = 0
...
+#        Default:     0 (disabled)
+#                     1 (enabled)
 #
 
 ItemDelete.Vendor = 0
 
 #
 #    ItemDelete.Quality
-#        Description:
+#        Description: Saving items into database that are Qualities greater than 3
```
Comment (FrancescoBorzi) [inline src/server/worldserver/worldserver.conf.dist]: please reword to `that have quality greater than ItemDelete.Quality`

File: src/server/worldserver/worldserver.conf.dist [hunk 5]
```diff
@@ -2970,23 +2970,35 @@ ItemDelete.Method = 0
...
+#
 #        Default:     3
 #
 
 ItemDelete.Quality = 3
 
 #
 #    ItemDelete.ItemLevel
-#        Description:
+#        Description: Saving items into database that are Item Levels greater than 80
```
Comment (FrancescoBorzi) [inline src/server/worldserver/worldserver.conf.dist]: `greater or equal to ItemDelete.ItemLevel`

File: src/server/game/Handlers/ItemHandler.cpp [hunk 6]
```diff
@@ -1687,3 +1700,25 @@ bool WorldSession::CanUseBank(uint64 bankerGUID) const
 
     return true;
 }
+
+bool WorldSession::recoveryItem(Item* pItem)
+{
+    bool method = sWorld->getBoolConfig(CONFIG_ITEMDELETE_METHOD);
+    uint32 quality = sWorld->getIntConfig(CONFIG_ITEMDELETE_QUALITY);
+    uint32 itemLevel = sWorld->getIntConfig(CONFIG_ITEMDELETE_ITEM_LEVEL);
```
Comment (FrancescoBorzi) [inline src/server/game/Handlers/ItemHandler.cpp]: why do we need to create new variables? we don't need to allocate extra memory

**Comment by FrancescoBorzi** [comment]
@IntelligentQuantum looks good :+1: 

Now waiting for a tester feedback.

**Comment by Helias** [comment]
I'm trying to test this PR but I got the following error during the import of db characters updates with latest master revision:

`ERROR 1062 (23000) at line 126: Duplicate entry '1572331570357004681' for key 'PRIMARY'`

Probably you need to generate again the sql pending file using `data/sql/updates/pending_db_characters/create_sql.sh` after updating this PR with master.

Anyway, I imported the table recovery_item manually but the feature does not work, I don't kno why.
Probably you added a condition that does not work properly.

**Comment by FrancescoBorzi** [comment]
@Helias no, we can't merge to master if it gives such an error. I've tried to put a different id (just re-generated). Can you please give another try?

Also @IntelligentQuantum just a tip for future: never push changes to your `master` branch, always create a new branch for each PR you want to open. But for this time it's fine.

**Comment by IntelligentQuantum** [comment]
@FrancescoBorzi Kk sorry
@Helias [Video](https://mega.nz/#!4zJHnYKY!XzuPkpsj1kGm41B2e42Up4Q_fcTSZWf9gbXM0mCBhX0)

**Comment by Helias** [comment]
I rebuild everything and checked the worldserver.conf, probably I missed something, now it works.

**Comment by BarbzYHOOL** [comment]
quite interesting, don't you think it will flood the DB though? I have seen the configuration options but still

**Comment by Nefertumm** [comment]
It would be awesome if it stores the unixtime of deletion in db, so we can implement an option for delete entries after X amount of days.

**Comment by mynew** [comment]
#0  0x0000000000c751ab in GetUInt32Value (index=<optimized out>, this=<optimized out>, index=<optimized out>) at ......../src/server/game/Entities/Object/Object.h:164
#1  GetEntry (this=<optimized out>) at ......../src/server/game/Entities/Object/Object.h:138
#2  WorldSession::HandleBuybackItem (this=<optimized out>, recvData=...) at ......../src/server/game/Handlers/ItemHandler.cpp:797
#3  0x0000000000ce55db in WorldSession::Update (this=0x7fff9d62e300, diff=<optimized out>, updater=...) at ......../src/server/game/Server/WorldSession.cpp:328
#4  0x0000000000c8ad81 in Map::Update (this=0x7fffc7f3a300, t_diff=0, s_diff=10) at ......../src/server/game/Maps/Map.cpp:672
#5  0x0000000000e44a99 in MapUpdateRequest::call (this=0x7fff31e40330) at ......../src/server/game/Maps/MapUpdater.cpp:57
#6  0x0000000000ea1f09 in DelayExecutor::svc (this=0x7ffff197cb68) at ......../src/common/Threading/DelayExecutor.cpp:53
#7  0x00007ffff74d56d7 in ACE_Task_Base::svc_run (args=0x7ffff197cb68) at Task.cpp:260
#8  0x00007ffff74d6dc5 in ACE_Thread_Adapter::invoke_i (this=<optimized out>) at Thread_Adapter.cpp:161
#9  0x00007ffff74d6d0f in ACE_Thread_Adapter::invoke (this=0x7fffbbb40d00) at Thread_Adapter.cpp:96
#10 0x00007ffff63e4184 in start_thread (arg=0x7fffb3cd2700) at pthread_create.c:312
#11 0x00007ffff56ed03d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111

Crash  ItemHandler.cpp
stmt->setUInt32(1, pItem->GetEntry());

**Comment by Nefertumm** [comment]
@mynew open a new issue please.

**Comment by GitVerge** [comment]
> #0 0x0000000000c751ab in GetUInt32Value (index=, this=, index=) at ......../src/server/game/Entities/Object/Object.h:164
> #1 GetEntry (this=) at ......../src/server/game/Entities/Object/Object.h:138
> #2 WorldSession::HandleBuybackItem (this=, recvData=...) at ......../src/server/game/Handlers/ItemHandler.cpp:797
> #3 0x0000000000ce55db in WorldSession::Update (this=0x7fff9d62e300, diff=, updater=...) at ......../src/server/game/Server/WorldSession.cpp:328
> #4 0x0000000000c8ad81 in Map::Update (this=0x7fffc7f3a300, t_diff=0, s_diff=10) at ......../src/server/game/Maps/Map.cpp:672
> #5 0x0000000000e44a99 in MapUpdateRequest::call (this=0x7fff31e40330) at ......../src/server/game/Maps/MapUpdater.cpp:57
> #6 0x0000000000ea1f09 in DelayExecutor::svc (this=0x7ffff197cb68) at ......../src/common/Threading/DelayExecutor.cpp:53
> #7 0x00007ffff74d56d7 in ACE_Task_Base::svc_run (args=0x7ffff197cb68) at Task.cpp:260
> #8 0x00007ffff74d6dc5 in ACE_Thread_Adapter::invoke_i (this=) at Thread_Adapter.cpp:161
> #9 0x00007ffff74d6d0f in ACE_Thread_Adapter::invoke (this=0x7fffbbb40d00) at Thread_Adapter.cpp:96
> #10 0x00007ffff63e4184 in start_thread (arg=0x7fffb3cd2700) at pthread_create.c:312
> #11 0x00007ffff56ed03d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
> 
> Crash ItemHandler.cpp
> stmt->setUInt32(1, pItem->GetEntry());

Just in case the issue won't be created, try moving all that new code on line ~797 to the beginning of the "if" so we avoid using pItem after calling StoreItem.

**Comment by Nefertumm** [comment]
That crash made me wonder, what if the seller bought back the item? He can claim the recovery but he has the item...

**Comment by GitVerge** [comment]
> That crash made me wonder, what if the seller bought back the item? He can claim the recovery but he has the item...

I think that the block of code that causes the crash is used to avoid that, it's deleting from the recovery table

**Comment by Nefertumm** [comment]
Oh yeah.
Also, what if the item has multiple entries? Recovering just one of them would delete all of them from the table.
This came to me just now, sorry :P

**Comment by IntelligentQuantum** [comment]
Hey guys

@BarbzYHOOL Actually we can flood db with CharacterRecovery && LagReport, btw we have configuration and ItemRecovery work like CharacterRecovery.

@GitVerge @Nefertumm tnx for help.

@mynew Can you test this diff please [Crash.diff](https://pastebin.com/DPTYePnK)

Sorry my english is bad! :smile:

**Pull request merged**